### PR TITLE
ST_OffsetCurve multiline + tests

### PR DESCRIPTION
**Pull request by Komzpa**
https://trac.osgeo.org/postgis/ticket/2508 now without topologyexception

### CODE REVIEW

File: liblwgeom/lwgeom_geos.h [hunk 1]
```diff
@@ -35,13 +35,24 @@ LWGEOM* GEOS2LWGEOM(const GEOSGeometry* geom, uint8_t want3d);
...
+GEOSGeometry* LWGEOM_GEOS_makeValid(const GEOSGeometry*);
 
 GEOSGeometry* make_geos_point(double x, double y);
 GEOSGeometry* make_geos_segment(double x1, double y1, double x2, double y2);
 
-int cluster_intersecting(GEOSGeometry** geoms, uint32_t num_geoms, GEOSGeometry*** clusterGeoms, uint32_t* num_clusters);
-int cluster_within_distance(LWGEOM** geoms, uint32_t num_geoms, double tolerance, LWGEOM*** clusterGeoms, uint32_t* num_clusters);
-int union_dbscan(LWGEOM** geoms, uint32_t num_geoms, UNIONFIND* uf, double eps, uint32_t min_points, char** is_in_cluster_ret);
+int
+cluster_intersecting(GEOSGeometry** geoms, uint32_t num_geoms, GEOSGeometry*** clusterGeoms, uint32_t* num_clusters);
```
Comment (dbaston) [inline liblwgeom/lwgeom_geos.h]: Are these related to the ticket?

File: liblwgeom/lwgeom_geos.c [hunk 2]
```diff
@@ -1206,8 +1210,11 @@ lwgeom_sharedpaths(const LWGEOM* geom1, const LWGEOM* geom2)
 	return result;
 }
 
-LWGEOM*
-lwgeom_offsetcurve(const LWLINE* lwline, double size, int quadsegs, int joinStyle, double mitreLimit)
+static LWGEOM*
```
Comment (dbaston) [inline liblwgeom/lwgeom_geos.c]: Consider removing this prototype

File: liblwgeom/lwgeom_geos.c [hunk 3]
```diff
@@ -1233,6 +1244,83 @@ lwgeom_offsetcurve(const LWLINE* lwline, double size, int quadsegs, int joinStyl
...
+	if (!lwgeom_is_collection(geom))
+		return lwgeom_offsetcurve_line(lwgeom_as_lwline(geom), size, quadsegs, joinStyle, mitreLimit);
+	else
+	{
+		LWCOLLECTION* result;
+		LWGEOM* tmp;
+		const LWCOLLECTION* col = lwgeom_as_lwcollection(geom);
+		int i;
+		result = lwcollection_construct_empty(MULTILINETYPE, srid, is3d, LW_FALSE);
+		for (i = 0; i < col->ngeoms; i++)
```
Comment (dbaston) [inline liblwgeom/lwgeom_geos.c]: `ngeoms` is unsigned, right?

File: liblwgeom/lwgeom_geos.c [hunk 4]
```diff
@@ -1233,6 +1244,83 @@ lwgeom_offsetcurve(const LWLINE* lwline, double size, int quadsegs, int joinStyl
...
+
+			if (lwgeom_get_type(tmp) != LINETYPE && lwgeom_get_type(tmp) != MULTILINETYPE) return NULL;
+
+			if (!lwgeom_is_empty(tmp))
+			{
+				if (!lwgeom_is_collection(tmp))
+					lwcollection_add_lwgeom(result, tmp);
+				else
+				{
+					const LWCOLLECTION* c2 = lwgeom_as_lwcollection(tmp);
```
Comment (dbaston) [inline liblwgeom/lwgeom_geos.c]: Consider a comment to explain what's happening in this block

File: liblwgeom/lwgeom_geos.c [hunk 5]
```diff
@@ -1233,6 +1244,83 @@ lwgeom_offsetcurve(const LWLINE* lwline, double size, int quadsegs, int joinStyl
...
+	{
+		LWGEOM* noded = lwgeom_node(geom);
+		if (noded)
+		{
+			result = lwgeom_offsetcurve_multilinestring(noded, size, quadsegs, joinStyle, mitreLimit);
+			lwfree(noded);
+			if (!result) lwerror("lwgeom_offsetcurve: noding input still does not let it be processed.");
+		}
+	}
+	tmp = lwgeom_make_valid(result);
```
Comment (dbaston) [inline liblwgeom/lwgeom_geos.c]: May need to adjust implicit makevalid based on closure of dev list discussion

File: liblwgeom/lwgeom_geos.c [hunk 6]
```diff
@@ -1206,8 +1210,11 @@ lwgeom_sharedpaths(const LWGEOM* geom1, const LWGEOM* geom2)
 	return result;
 }
 
-LWGEOM*
-lwgeom_offsetcurve(const LWLINE* lwline, double size, int quadsegs, int joinStyle, double mitreLimit)
+static LWGEOM*
+lwgeom_offsetcurve_line(const LWLINE* lwline, double size, int quadsegs, int joinStyle, double mitreLimit);
+
+static LWGEOM*
+lwgeom_offsetcurve_line(const LWLINE* lwline, double size, int quadsegs, int joinStyle, double mitreLimit)
```
Comment (dbaston) [inline liblwgeom/lwgeom_geos.c]: `lwline_offsetcurve` and `lwmline_offsetcurve` would more closely follow existing conventions, IMO

File: liblwgeom/lwgeom_geos_clean.c [hunk 7]
```diff
@@ -321,34 +321,37 @@ static GEOSGeometry*
...
-	if (!point) return NULL;
+	/* if noding isn't enough, fall back to self-union */
+	if (!GEOSisValid(noded))
+	{
+		GEOSGeometry* point;
+		GEOSGeometry* unioned;
 
-	LWDEBUGF(3, "Boundary point: %s", lwgeom_to_ewkt(GEOS2LWGEOM(point, 0)));
+		/*
+		 * Union with first geometry point, obtaining full noding
```
Comment (dbaston) [inline liblwgeom/lwgeom_geos_clean.c]: Could this be replaced with a call to UnaryUnion?

File: liblwgeom/lwgeom_geos.c (hunk 3 above)
Comment (Komzpa) [reply liblwgeom/lwgeom_geos.c]: @Algunenano how do you get the integer warnings? I miss them on travis :)

Comment (Algunenano) [reply liblwgeom/lwgeom_geos.c]: I have a pretty big list of flags, but you are probably looking for `-Wconversion`. Right now it leads to a lot of warnings, but hopefully I'll address most of them in https://github.com/postgis/postgis/pull/208

File: liblwgeom/lwgeom_geos.c (hunk 4 above)
Comment (Komzpa) [reply liblwgeom/lwgeom_geos.c]: made separate function for collection inline concatenation

File: liblwgeom/lwgeom_geos.c (hunk 5 above)
Comment (Komzpa) [reply liblwgeom/lwgeom_geos.c]: removed makevalid for now

File: liblwgeom/lwgeom_geos.c (hunk 2 above)
Comment (Komzpa) [reply liblwgeom/lwgeom_geos.c]: done

File: liblwgeom/lwgeom_geos.h (hunk 1 above)
Comment (Komzpa) [reply liblwgeom/lwgeom_geos.h]: removed

File: liblwgeom/lwgeom_geos.c (hunk 3 above)
Comment (Komzpa) [reply liblwgeom/lwgeom_geos.c]: @dbaston fixed

**Comment by Komzpa** [comment]
@dbaston @Algunenano can you wave if it's ok to commit?

File: liblwgeom/lwgeom_geos.c [hunk 8]
```diff
@@ -1223,13 +1227,97 @@ lwgeom_offsetcurve(const LWLINE* lwline, double size, int quadsegs, int joinStyl
...
+}
+
+static LWGEOM*
+lwmline_offsetcurve(const LWGEOM* geom, double size, int quadsegs, int joinStyle, double mitreLimit)
+{
+	int32_t srid = get_result_srid(geom, NULL, __func__);
+	uint8_t is3d = FLAGS_GET_Z(geom->flags);
+	if (srid == SRID_INVALID) return NULL;
+
+	if (geom->type == LINETYPE)
```
Comment (Algunenano) [inline liblwgeom/lwgeom_geos.c]: The rest of the codebase uses case statements instead of if-defs. Please, consider switching it

File: liblwgeom/lwgeom_geos.c [hunk 9]
```diff
@@ -1223,13 +1227,97 @@ lwgeom_offsetcurve(const LWLINE* lwline, double size, int quadsegs, int joinStyl
...
+
+	if (lwgeom_dimension(geom) != 1)
+	{
+		lwerror("%s: input is not linear", __func__, lwtype_name(geom->type));
+		return NULL;
+	}
+
+	result = lwmline_offsetcurve(geom, size, quadsegs, joinStyle, mitreLimit);
+	if (!result)
+	{
```
Comment (Algunenano) [inline liblwgeom/lwgeom_geos.c]: Is this going to print the `lwerror("%s: unsupported geometry type: %s", __func__, lwtype_name(geom->type));` from above?

File: liblwgeom/lwgeom_geos.c (hunk 8 above)
Comment (Komzpa) [reply liblwgeom/lwgeom_geos.c]: switched to case.

File: liblwgeom/lwgeom_geos.c (hunk 9 above)
Comment (Komzpa) [reply liblwgeom/lwgeom_geos.c]: in context of postgres, lwerror is pg_error https://github.com/postgis/postgis/blob/c7e346cb4b4f04737c50f469b170ed288ab7c3b7/libpgcommon/lwgeom_pg.c#L121 which in turn calls `ereport(ERROR` which in turn cancels query at the spot (see https://www.postgresql.org/docs/10/static/error-message-reporting.html) - so if it prints that, it won't get to this stage.

File: liblwgeom/lwcollection.c [hunk 10]
```diff
@@ -214,24 +214,32 @@ LWCOLLECTION* lwcollection_add_lwgeom(LWCOLLECTION *col, const LWGEOM *geom)
...
-	}
 #endif
 
 	col->geoms[col->ngeoms] = (LWGEOM*)geom;
 	col->ngeoms++;
 	return col;
 }
 
+LWCOLLECTION*
+lwcollection_concat_in_place(LWCOLLECTION* col1, const LWCOLLECTION* col2)
```
Comment (dbaston) [inline liblwgeom/lwcollection.c]: Add source comment to document what this function does, and what responsibilities of caller are. (It looks like caller would need to release col2 after calling?)

File: liblwgeom/lwgeom_geos.c (hunk 6 above)
Comment (dbaston) [reply liblwgeom/lwgeom_geos.c]: Sorry, maybe this comment was unclear. The convention would be to have a function `lwgeom_offsetcurve` that takes an argument `LWGEOM` and uses the `switch` statement to dispatch to `lwline_offsetcurve`, `lwmline_offsetcurve`, `lwcollection_offsetcurve`, etc. as needed. They type-specific variants can all be public. See for example `lwgeom_segmentize2d` in the header at line 1481.
